### PR TITLE
4391 by Inlead: Place2book: Default values are overriding given values.

### PR DESCRIPTION
**Pull request by NicolaeGrigoriu**
#### Link to issue

https://platform.dandigbib.org/issues/4391

#### Description

Once you have created an event and defined values in Maintain copy, Kultunaut export and Passive event as you wish, then it stores the event fine and correct. But if you edit the event, it sets the default values again even if you do not want it to change

**Example 1:**

I have in the standards for p2b under /admin/config/ding/place2book/defaults checked Maintain copy, Kultunaut export and Passive arrangement.

I create an event that I want there to be ticket to. So I remove the checkmark in "Passive".

I save the event.

I would like to edit the body-field of the event, so I edit the event.

I now notice that "Passive" is again marked in the standards.

If I accidentally save, the ticket link is gone, and I cannot undo to a non-passive event, and if tickets have been sold, I cannot delete and start over.

**Example 2:**

I have in the standards for p2b under /admin/config/ding/place2book/defaults checked Maintain copy, and Kultunaut export

I want to make an event where there should not be tickets nor sent to Kultunaut. So I remove the checkmark in Maintain copy and Kultunaut export.

I save the event.

I would like to edit the body-field of the event, so I edit the event.

I now notice that Maintenance copy and Kultunaut export are checked again by the standards.

This means that a ticket will be created for an event, which should not be ticketed and also sent incorrectly to Kultunaut.

**Solution:** 

As mentioned in the API documentation in the PUT request the header input Content-Type should be application/json. Plus, get passive value direct from the event entity.

Could not reproduce example 2.

#### Screenshot of the result

image

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

### CODE REVIEW

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 44138889eff049feb2e7e3b224198559d31f2aa5
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 44138889eff049feb2e7e3b224198559d31f2aa5
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by holt83** [comment]
Why not use the passive boolean returned from the API instead of adding af new field?

File: modules/ding_place2book/ding_place2book.fields.inc [hunk 1]
```diff
@@ -276,12 +276,14 @@ function ding_place2book_field_widget_form(&$form, &$form_state, $field, $instan
         $element['place2book']['prices_wrapper']['prices'][$key]['sale_begin_at'] = array(
           '#type' => 'date_popup',
           '#default_value' => $sale_begin_at,
+          '#states' => $required_state,
```
Comment (holt83) [inline modules/ding_place2book/ding_place2book.fields.inc]: Seems unrelated?

File: modules/ding_place2book/lib/p2b/P2b.php [hunk 2]
```diff
@@ -559,7 +559,7 @@ private function p2bRequest($url, $type, array $params = array(), $code = 200) {
     );
 
     if (!empty($params)) {
-      if ($type == 'POST') {
+      if ($type == 'POST' || $type == 'PUT') {
```
Comment (holt83) [inline modules/ding_place2book/lib/p2b/P2b.php]: Seems unrelated?

Comment (NicolaeGrigoriu) [reply modules/ding_place2book/lib/p2b/P2b.php]: As mentioned in the documentation http://developer.place2book.com/v1/events/#update-an-event. Header options should be of the type json

Comment (holt83) [reply modules/ding_place2book/lib/p2b/P2b.php]: Sorry, I just can see how this solves the problem of default settings overriding actual values?

File: modules/ding_place2book/ding_place2book.fields.inc (hunk 1 above)
Comment (holt83) [reply modules/ding_place2book/ding_place2book.fields.inc]: Thanks for fix 👍

File: modules/ding_place2book/lib/p2b/P2b.php (hunk 2 above)
Comment (NicolaeGrigoriu) [reply modules/ding_place2book/lib/p2b/P2b.php]: when the node form is saved for the first time the POST request is sent to the p2b server with the `passive` property in a json object. This property is received on GET request and assigend in the line 180(https://github.com/ding2/ding2/pull/1472/files#diff-8f3bc34f993926671354976be58fae7aR180). 
But when the node is updated the PUT method is used. If no Content-Type json is set then it seems that an empty object is sent. So, the entity is not updated on the p2b side.

Comment (holt83) [reply modules/ding_place2book/lib/p2b/P2b.php]: Ok, thanks for elaborating. So the problem wasn't really defaults overriding, but that the event was never updated?

When looking at how we determine the passive state now it just looks weird:

```
      $passive = $event->passive ?? 0;
      if (!isset($meta_data->event_state)) {
        $passive = variable_get('ding_p2b_default_passive', FALSE);
      }
      elseif ($meta_data->event_state == 'passive') {
        $passive = 1;
      }
```

So it looks like passive state can be found two places and that the one on$meta_data->event_state takes precedence. And if the there's nothing set on $meta_data->event_state it always fallback to default value regardless of the value on @event->passive.

Is this really correct? I would think one of them would be the correct one to get the passive state.

Comment (NicolaeGrigoriu) [reply modules/ding_place2book/lib/p2b/P2b.php]: It seems that the problem is not in the module but in the place2book service. [See video](https://www.dropbox.com/s/utreqbku41wg87h/p2b.mp4?dl=0). When event is firstly created it saves all the needed fields. But when a price is created for a specific event then it recives the published status.

Comment (NicolaeGrigoriu) [reply modules/ding_place2book/lib/p2b/P2b.php]: One more thing, when `$event->passive == TRUE` then `$meta_data->event_state == 'passive'` and `$event->state == 'passive'`.

Comment (holt83) [reply modules/ding_place2book/lib/p2b/P2b.php]: > One more thing, when `$event->passive == TRUE` then `$meta_data->event_state == 'passive'` and `$event->state == 'passive'`.

Ok. Why not just check one of them then? It seems excessive and needlessly complex to check them all.

Comment (holt83) [reply modules/ding_place2book/lib/p2b/P2b.php]: > It seems that the problem is not in the module but in the place2book service. [See video](https://www.dropbox.com/s/utreqbku41wg87h/p2b.mp4?dl=0). When event is firstly created it saves all the needed fields. But when a price is created for a specific event then it receives the published status. Is this on purpose?

Thanks for the vid. Really helpful 👍 

I guess it's because it doesn't make sense to have tickets on a passive event, so p2b automatically publishes the event. So yeah, I think it's on purpose.

Comment (NicolaeGrigoriu) [reply modules/ding_place2book/lib/p2b/P2b.php]: If so, then the fix for the node update should be enough, I guess :) Everything else was reverted to initial state.
